From a743d744a85778b2cceb1d1c2824f1a672d81e88 Mon Sep 17 00:00:00 2001
From: Julian Greilich <j.greilich@gmx.de>
Date: Tue, 28 Oct 2025 11:30:35 +0100
Subject: a11y iOS: Implement accessibilityHitTest for QUIView

This is to respect the correct paint order of the accessible
elements.

This function has only an effect for iOS 18 and higher.

Task-number: QTBUG-140393
Change-Id: I3fcdde9db432dd7e6f69c98001b12fcf1c85eb87
---
 .../platforms/ios/quiview_accessibility.mm    | 28 +++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git x/qtbase/src/plugins/platforms/ios/quiview_accessibility.mm y/qtbase/src/plugins/platforms/ios/quiview_accessibility.mm
index da8a0c4df904b63c07afb054d4445a9783153bbf..ecc48e3818b4591dee9d0667ac1a47a16ff96695 100644
--- x/qtbase/src/plugins/platforms/ios/quiview_accessibility.mm
+++ y/qtbase/src/plugins/platforms/ios/quiview_accessibility.mm
@@ -55,6 +55,34 @@
     [m_accessibleElements removeAllObjects];
 }
 
+- (id)accessibilityHitTest:(CGPoint)point withEvent:(UIEvent *)event
+{
+    for (UIView *subview in [self.subviews reverseObjectEnumerator]) {
+        if (subview.hidden || qFuzzyIsNull(subview.alpha))
+            continue;
+
+        CGPoint viewPoint = [subview convertPoint:point fromView:self];
+        if (id child = [subview accessibilityHitTest:viewPoint withEvent:event])
+            return child;
+    }
+
+    if (!self.platformWindow)
+        return [super accessibilityHitTest:point withEvent:event];
+
+    QWindow *win = self.platformWindow->window();
+    QAccessibleInterface *iface = win->accessibleRoot();
+    if (!iface)
+        return nil;
+
+    QPoint screenPoint = self.platformWindow->mapToGlobal(QPointF::fromCGPoint(point).toPoint());
+    QAccessibleInterface *child = iface->childAt(screenPoint.x(), screenPoint.y());
+    if (!child || !child->state().focusable)
+        return nil;
+
+    QAccessible::Id childId = QAccessible::uniqueId(child);
+    return [QMacAccessibilityElement elementWithId:childId];
+}
+
 // this is a container, returning yes here means the functions below will never be called
 - (BOOL)isAccessibilityElement
 {
